pt/lessons/advanced/metaprogramming.md

```diff
@@ -0,0 +1,246 @@
+---
+layout: page
+title: Metaprogramação
+category: advanced
+order: 7
+lang: pt
+---
+
+Metaprogramação é o processo de utilização de código para escrever código. Em Elixir isso nos dá a capacidade de estender a linguagem para se adequar às nossas necessidades e dinamicamente alterar o código. Vamos começar observando como Elixir é representado por debaixo dos panos, em seguida como modificá-lo, e finalmente, como podemos usar esse conhecimento para estendê-lo.
+
+
+Uma palavra de cautela: Metaprogramação é complicado e só deve ser usado quando for absolutamente necessário. O uso excessivo certamente levará a um código complexo que é difícil de entender e debugar.
+
+## Sumário
+
+- [Quote](#quote)
+- [Unquote](#unquote)
+- [Macros](#macros)
+	- [Macros Privados](#macros-privados)
+	- [Higienização de Macros](#higienizacao-de-macros)
+	- [Binding](#binding)
+
+## Quote
+
+O primeiro passo para metaprogramação é a compreensão de como as expressões são representadas. Em Elixir, a árvore de sintaxe abstrata (AST), a representação interna do nosso código, é composta de tuplas. Estas tuplas contêm três partes: o nome da função, metadados e argumentos da função.
+
+A fim de ver essas estruturas internas, Elixir fornece-nos a função `quote/2`. Usando `quote/2` podemos converter o código Elixir em sua representação subjacente:
+
+```elixir
+iex> quote do: 42
+42
+iex> quote do: "Hello"
+"Hello"
+iex> quote do: :world
+:world
+iex> quote do: 1 + 2
+{:+, [context: Elixir, import: Kernel], [1, 2]}
+iex> quote do: if value, do: "True", else: "False"
+{:if, [context: Elixir, import: Kernel],
+ [{:value, [], Elixir}, [do: "True", else: "False"]]}
+iex(6)>
+```
+
+Notou que os três primeiros exemplos não retornaram tuplas? Existe cinco literais que retornam ele mesmo quando citados (quoted):
+
+```elixir
+iex> :atom
+:atom
+iex> "string"
+"string"
+iex> 1 # All numbers
+1
+iex> [1, 2] # Lists
+[1, 2]
+iex> {"hello", :world} # 2 element tuples
+{"hello", :world}
+```
+
+## Unquote
+
+Agora que podemos recuperar a estrutura interna do nosso código, como podemos modificá-lo? Para injetar novo código ou valores usamos `unquote/1`. Quando *unquote* em uma expressão que vai ser avaliada e injetado no AST. Para demonstrar `unquote/1` vejamos alguns exemplos:
+
+```elixir
+iex> denominator = 2
+2
+iex> quote do: divide(42, denominator)
+{:divide, [], [42, {:denominator, [], Elixir}]}
+iex> quote do: divide(42, unquote(denominator))
+{:divide, [], [42, 2]}
+```
+
+No primeiro exemplo, a variável `denominator` é citado de modo que AST resultante inclua um tuplo para aceder a variável. No `unquote/1` exemplo, o código resultante inclui o valor de `denominator` no lugar.
```

Comment:
maybe: "a variável `denominator` é citada de modo que a AST resultante inclua uma tupla para acessar a variável."

and: "No exemplo de `unquote/1` o código ..."